Rewrite PodSecurityPolicy migration guide (#31782)
kubernetes/website: content/en/docs/tasks/configure-pod-container/migrate-from-psp.md

- Ensure the `PodSecurity` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/#feature-gates-for-alpha-or-beta-features) is enabled.

This page assumes you are already familiar with the basic [Pod Security
Admission](/docs/concepts/security/pod-security-admission/) concepts.

<!-- body -->

> ## Overview

Reviewer: ```suggestion
## Overall approach
```

(it's sort of an anti-pattern to have the page overview and then use Overview as the heading of the first subsequent section)

  Security Admission does expose some [static configuration for
  exemptions](/docs/concepts/security/pod-security-admission/#exemptions).

Even if Pod Security Admission does not meet all of your needs it was designed to be _complementary_
to other policy enforcement mechanisms, and can provide a useful fallback running alongside other
admission webhooks.

[dynamic-admission]: /docs/reference/access-authn-authz/extensible-admission-controllers/

## 1. Review namespace permissions

Pod Security Admission is controlled by [labels on
namespaces](/docs/concepts/security/pod-security-admission/#pod-security-admission-labels-for-namespaces).
This means that anyone who can update (or patch or create) a namespace can also modify the Pod
Security level for that namespace, which could be used to bypass a more restrictive policy. Before
proceeding, ensure that only trusted, privileged users have these namespace permissions. It is not
recommended to grant these powerful permissions to users that shouldn't have elevated permissions,
but if you must you will need to use an [admission webhook][dynamic-admission] to place additional
restrictions on Pod Security labels.

> ## 2. Simplify & Standardize PodSecurityPolicies {#simplify-psps}

Reviewer: (nit)
```suggestion
## 2. Simplify & standardize PodSecurityPolicies {#simplify-psps}
```

In this section, you will reduce mutating PodSecurityPolicies and remove options that are outside
the scope of the PodSecurityStandards. You should make the changes recommended here to an offline
copy of the original PodSecurityPolicy being modified. The cloned profile should have a different
name that is alphabetically before the original (for example, prepend a `0` to it). Do not create the
new policies in Kubernetes yet - that will be covered in the [Rollout the updated
policies](#psp-update-rollout) section below.

> ### 2.a. Eliminate purely mutating fields

Reviewer: This heading doesn't feel quite right - do you have any other wording you considered here?

Reviewer: Let's live with it. I can't suggest something better.

validating policy. These fields (also listed in the [table below](#psp-fields)) are:

- `.spec.defaultAllowPrivilegeEscalation`
- `.spec.runtimeClass.defaultRuntimeClassName`
- `.metadata.annotations['seccomp.security.alpha.kubernetes.io/defaultProfileName']`
- `.metadata.annotations['apparmor.security.beta.kubernetes.io/defaultProfileName']`
- `.spec.defaultAddCapabilities` - Although technically a mutating & validating field, these should
  be merged into `.spec.allowedCapabilities` which performs the same validation without mutation.

{{< caution >}}
Removing these could result in workloads missing required configuration, and cause problems. See
[Rollout the updated policies](#psp-update-rollout) below for advice on how to roll these changes
out safely.
{{< /caution >}}

### 2.b. Eliminate options not covered by the Pod Security Standards

There are several fields in PodSecurityPolicy that are not covered by the Pod Security Standards. If
you must enforce these options, you will need to supplement Pod Security Admission with an
[admission webhook][dynamic-admission], which is outside the scope of this guide.

> First, you can remove the purely validating fields that Pod Security does not cover. These fields

Reviewer 2: What does "Pod Security" stand for here? The "Pod Security Admission" or "PodSecurityPolicy"?

(also listed in the [table below](#psp-fields) with "no opinion") are:

- `.spec.allowedHostPaths`
- `.spec.allowedFlexVolumes`
- `.spec.allowedCSIDrivers`
- `.spec.forbiddenSysctls`
- `.spec.runtimeClass`

> Additionally, the following unix group related fields can also be removed.

Reviewer: My suggestion:
```
You can also remove the following fields, that are related to POSIX / UNIX group controls.
```

{{< caution >}}
If any of these use the `MustRunAs` strategy they may be mutating! Removing these could result in
workloads not setting the required groups, and cause problems. See [Rollout the updated
policies](#psp-update-rollout) below for advice on how to roll these changes out safely.
{{< /caution >}}

- `.spec.runAsGroup`
- `.spec.supplementalGroups`
- `.spec.fsGroup`

The remaining mutating fields are required to properly support the Pod Security Standards, and will
need to be handled on a case-by-case basis later:

- `.spec.requiredDropCapabilities` - Required to drop `ALL` for the restricted profile.
- `.spec.seLinux` - (Only mutating with the `MustRunAs` rule) required to enforce the SELinux
  requirements of the baseline & restricted profiles.

> - `.spec.runAsUser` - (Non-mutating with the `RunAsAny` rule) required to enforce RunAsNonRoot for

Reviewer 2: ```suggestion
- `.spec.runAsUser` - (Non-mutating with the `RunAsAny` rule) required to enforce `RunAsNonRoot` for
```

   identified, the original Pod or PodTemplate should be updated with the desired configuration.
   As a reminder, the fields to review are:
   - `.spec.containers[*].securityContext.allowPrivilegeEscalation`
   - `.spec.initContainers[*].securityContext.allowPrivilegeEscalation`
   - `.spec.runtimeClassName`
   - `.spec.securityContext.seccompProfile`
   - `.metadata.annotations['container.apparmor.security.beta.kubernetes.io/*']` (replace * with each container name)
   - On containers, under `.spec.containers[*]` and `.spec.initContainers[*]`:
       - `.securityContext.allowPrivilegeEscalation`
       - `.securityContext.seccompProfile`
       - `.securityContext.capabilities.add`
3. Create the new PodSecurityPolicies. If any Roles or ClusterRoles are granting `use` on all PSPs
   this could cause the new PSPs to be used instead of their mutating counter-parts.
4. Update your authorization to grant access to the new PSPs. In RBAC this means updating any Roles
   or ClusterRoles that grant the `use` permision on the original PSP to also grant it to the
   updated PSP.
5. Verify: after some soak time, rerun the command from step 1 to see if any pods are still using

>    the original profiles. If the pods haven't been updated since the new policies were rolled out,

Reviewer 2: ```suggestion
   the original PodSecurityPolicies. If the pods haven't been updated since the new policies were rolled out,
```

   they may need to be cycled to be verified against the new PSPs.
6. (optional) Once you have verified that the original PSPs are no longer in use, you can delete
   them.

## 3. Update Namespaces

The following steps will need to be performed on every namespace in the cluster. Commands referenced
in these steps use the `$NAMESPACE` variable to refer to the namespace being updated.

### 3.a. Identify an appropriate Pod Security level

Start reviewing the [Pod Security Standards](/docs/concepts/security/pod-security-standards/) and

> familiarizing yourself with the 3 different profile levels.

Reviewer 2: ```suggestion
familiarizing yourself with the 3 different Pod security levels.
```

There are several ways to choose a Pod Security level for your namespace:

> 1. **By security requirements for the namespace** - if you are familiar with the expected access

Reviewer 2: ```suggestion
1. **By security requirements for the namespace** - If you are familiar with the expected access
```

   level for the namespace, you can choose an appropriate level based on those requirements, similar
   to how one might approach this on a new cluster.
2. **By existing Pod Security Policies** - Using the [table below](#psp-fields) you can map each

>    PSP to a Pod Security Standard level. If your PSPs aren't based on the PodSecurityStandards, you

Reviewer 2: ```suggestion
   PSP to a Pod Security level. If your PSPs aren't based on the PodSecurityStandards, you
```

   may need to decide between choosing a level that is at least as permissive as the PSP, and a
   level that is at least as restrictive. You can see which PSPs are in use for pods in a given
   namespace with this command:
   ```sh
   kubectl get pods -n $NAMESPACE -o jsonpath="{.items[*].metadata.annotations.kubernetes\.io\/psp}" | tr " " "\n" | sort -u
   ```
3. **By existing pods** - Using the strategies under [Verify the Pod Security

>    level](#verify-pss-level), you can test out both the Baseline and Restricted levels to see

Reviewer 2: Do we want to use 'Baseline' or 'baseline'? Please make them consistent through out the page.

kubectl label --dry-run=server --overwrite ns $NAMESPACE pod-security.kubernetes.io/enforce=$LEVEL
```
This command will return a warning for any _existing_ pods that are not valid under the proposed
level.

The second option is better for catching workloads that are not currently running: audit mode. When
running under audit-mode (as opposed to enforcing), pods that violate the policy level are recorded
in the audit logs, which can be reviewed later after some soak time. Warning mode works similarly,
but returns the warning to the user immediately. You can set the audit level on a namespace with
this command:
```sh
kubectl label --overwrite ns $NAMESPACE pod-security.kubernetes.io/audit=$LEVEL
```

If either of these approaches yield unexpected violations, you will need to either update the
violating workloads to meet the policy requirements, or relax the namespace Pod Security level.

### 3.c. Enforce the Pod Security level

When you are satisfied that the chosen level can safely be enforced on the namespace, you can update
the namespace to enforce the desired level:

```sh
kubectl label --overwrite ns $NAMESPACE pod-security.kubernetes.io/enforce=$LEVEL
```

> ### 3.d. "Disable" PodSecurityPolicy

Reviewer: ```suggestion
### 3.d. Bypass PodSecurityPolicy for that namespace
```

Finally, you can effectively "disable" PodSecurityPolicy at the namespace level by binding the fully
{{< example file="policy/privileged-psp.yaml" >}}privileged PSP{{< /example >}} to all service
accounts in the namespace.

```sh
# The following cluster-scoped commands are only needed once.
kubectl apply -f privileged-psp.yaml
kubectl create clusterrole privileged-psp --verb use --resource podsecuritypolicies.policy --resource-name privileged

# Per-namespace disable
kubectl create -n $NAMESPACE rolebinding disable-psp --clusterrole privileged-psp --group system:serviceaccounts:$NAMESPACE
```

> Since the privileged service account is non-mutating, and the PSP admission controller always

Reviewer 2: This is the first time we are talking about "privileged service account" ... why service account is relevant here?

prefers non-mutating PSPs, this will ensure that pods in this namespace are no longer being modified
or restricted by PodSecurityPolicy.

The advantage to disabling PodSecurityPolicy on a per-namespace basis like this is if a problem

> arises you can easily roll the change back by deleting the rolebinding. Just make sure the

Reviewer 2: ```suggestion
arises you can easily roll the change back by deleting the RoleBinding. Just make sure the
```

I0218 00:59:44.903350      13 plugins.go:161] Loaded 14 validating admission controller(s) successfully in the following order: LimitRanger,ServiceAccount,PodSecurity,Priority,PersistentVolumeClaimResize,RuntimeClass,CertificateApproval,CertificateSigning,CertificateSubjectRestriction,DenyServiceExternalIPs,ValidatingAdmissionWebhook,ResourceQuota.
```

You should see `PodSecurity` (in the validating admission controllers), and neither list should
contain `PodSecurityPolicy`.

Once you are certain the PSP admission controller is disabled (and after sufficient soak time to be
confident you won't need to roll back), you are free to delete your PodSecurityPolicies and any
associated Roles, ClusterRoles, RoleBindings and ClusterRoleBindings (just make sure they don't
grant any other unrelated permissions).

> ## Appendix: Pod Security Policies to Pod Security Standards {#psp-fields}

Reviewer: I would put this section / table as a standalone page in the Reference section, and hyperlink to it.

(Appendices to tasks can be harder to find than standalone reference pages, because maybe the task you're performing is “find out what Pod Security Standard I should use for this namespace, the vendor's docs only mention a PodSecurityPolicy” and we don't have a task page for that).

Author: Sounds good, I'll follow up with this.

The following table enumerates the fields in the `PodSecurityPolicySpec` API, along with the allowed
values for the [Baseline](/docs/concepts/security/pod-security-standards/#baseline) and
[Restricted](/docs/concepts/security/pod-security-standards/#restricted) profiles. Anything outside
the allowed values for those profiles would fall under the
[Privileged](/docs/concepts/security/pod-security-standards/#priveleged) profile.

<table>

> 	<caption style="display:none"></caption>

Reviewer: For accessibility, we should set a caption.

Author: Oops, thanks.

	<tbody>
		<tr>
      <th style="white-space: nowrap;"><code>PodSecurityPolicy</code></th>
      <th>Type</th>

>       <th>PodSecurity Equivalent</th>

Reviewer 2: By "PodSecurity", are we referring to "PodSecurityStandard" or "PodSecurityPolicy"?

Author: Updated to `Pod Security Standards`

    <tr>
      <td style="white-space: nowrap;"><code>hostIPC</code></td>
      <td>Validating</td>
      <td><b>Baseline & Restricted</b>: <code>false</code> / undefined / nil</td>
    </tr>
    <tr>
      <td style="white-space: nowrap;"><code>seLinux</code></td>
      <td>Mutating & Validating</td>
      <td>
        <p><b>Baseline & Restricted</b>:
        <code>seLinux.rule</code> is <code>MustRunAs</code>, with the following <code>options</code></p>
        <ul>
          <li><code>user</code> is unset (<code>""</code> / undefined / nil)</li>
          <li><code>role</code> is unset (<code>""</code> / undefined / nil)</li>
          <li><code>type</code> is unset or one of: <code>container_t, container_init_t, container_kvm_t</code></li>
          <li><code>level</code> is anything</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td style="white-space: nowrap;"><code>runAsUser</code></td>
      <td>Mutating & Validating</td>
      <td>
        <p><b>Baseline</b>: Anything</p>
        <p><b>Restricted</b>: <code>rule</code> is <code>MustRunAsNonRoot</code> OR
        <code>MustRunAs</code> with <code>ranges</code> set to a non-empty list that doesn't include 0.</p>

>         <p><i>Note: Prior to v1.23, only the <code>MustRunAsNonRoot</code> rule is compatible with the Restricted profile.</i></p>

Author: Yes, thanks. I forgot we punted on allowing it to be unset when `runAsUser` was non-zero.

      <td>Validating</td>
      <td><i>No opinion (volumes takes precedence)</i></td>
    </tr>
    <tr>
      <td style="white-space: nowrap;"><code>allowedUnsafeSysctls</code></td>
      <td>Validating</td>
      <td><b>Baseline & Restricted</b>: undefined / nil / empty</td>
    </tr>
    <tr>
      <td style="white-space: nowrap;"><code>forbiddenSysctls</code></td>
      <td>Validating</td>
      <td><i>No opinion</i></td>
    </tr>
    <tr>
      <td style="white-space: nowrap;"><code>allowedProcMountTypes</code><br><i>(alpha feature)</i></td>
      <td>Validating</td>
      <td><b>Baseline & Restricted</b>: <code>["Default"]</code> OR undefined / nil / empty</td>
    </tr>
    <tr><td colspan="3"><code>runtimeClass</code></td></tr>
    <tr>
      <td style="white-space: nowrap;"><code>&nbsp;.defaultRuntimeClassName</code></td>
      <td>Mutating</td>
      <td><i>No opinion</i></td>
    </tr>
    <tr>
      <td style="white-space: nowrap;"><code>&nbsp;.allowedRuntimeClassNames</code></td>
      <td>Validating</td>
      <td><i>No opinion</i></td>
    </tr>
    <tr><td colspan="3"><b><code>metadata.annotations</code></b></td></tr>
    <tr>

>       <td colspan="3"><code>seccomp.security.alpha.kubernetes.io</code></td>

Author: Yeah. I did that because otherwise the column gets too wide and messes with the rest of the table formatting... I could cram it into each cell with a line break, would that be more clear? Can I leave the metadata.annototions bit on its own line?

The other option is to put it in a separate table to allow the columns to be sized differently.

>       <td colspan="3"><code>seccomp.security.alpha.kubernetes.io</code></td>

Reviewer: If you want the table to have a minimum width, we're open to setting a `class` on the table, or the row and having some accompanying CSS changes (ideally, in a separate commit).

I'd also prefer to have the white space style nowrap done with a class and some CSS.

    </tr>
    <tr>
      <td style="white-space: nowrap;"><code>&nbsp;/defaultProfileName</code></td>
      <td>Mutating</td>
      <td><i>No opinion</i></td>
    </tr>
    <tr>
      <td style="white-space: nowrap;"><code>&nbsp;/allowedProfileNames</code></td>
      <td>Validating</td>
      <td>
        <p><b>Baseline</b>: <code>"runtime/default,"</code> <i>(Trailing comma to allow unset)</i></p>
        <p><b>Restricted</b>: <code>"runtime/default"</code> <i>(No trailing comma)</i></p>
        <p><i><code>localhost/*</code> values are also permitted for both baseline & restricted.</i></p>
      </td>
    </tr>
    <tr>
      <td colspan="3"><code>apparmor.security.beta.kubernetes.io</code></td>
    </tr>
    <tr>
      <td style="white-space: nowrap;"><code>&nbsp;/defaultProfileName</code></td>
      <td>Mutating</td>
      <td><i>No opinion</i></td>
    </tr>
    <tr>
      <td style="white-space: nowrap;"><code>&nbsp;/allowedProfileNames</code></td>
      <td>Validating</td>
      <td>
        <p><b>Baseline</b>: <code>"runtime/default,"</code> <i>(Trailing comma to allow unset)</i></p>

>         <p><b>Restricted</b>: <code>"runtime/default,"</code> <i>(No trailing comma)</i></p>

Author: Whoops! Good catch.

- Ensure the `PodSecurity` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/#feature-gates-for-alpha-or-beta-features) is enabled.

This page assumes you are already familiar with the basic [Pod Security Admission](/docs/concepts/security/pod-security-admission/)
concepts.

<!-- body -->

## Overall approach

There are multiple strategies you can take for migrating from PodSecurityPolicy to Pod Security
Admission. The following steps are one possible migration path, with a goal of minimizing both the
risks of a production outage and of a security gap.

<!-- Keep section header numbering in sync with this list. -->
0. Decide whether Pod Security Admission is the right fit for your use case.
1. Review namespace permissions
2. Simplify & standardize PodSecurityPolicies
3. Update namespaces
   1. Identify an appropriate Pod Security level
   2. Verify the Pod Security level
   3. Enforce the Pod Security level
   4. Bypass PodSecurityPolicy
4. Review namespace creation processes
5. Disable PodSecurityPolicy

> ## 0. Decide whether Pod Security Admission is right for you

Reviewer: For all of these numbered headers, I would change them to be like:
```markdown
## 0. Decide whether Pod Security Admission is right for you {#decide-whether-pod-security-admission-is-right-for-you}
```
which _omits_ the numbering. That then gives it a fragment identifier which is stable if we renumber.

(It also leaves room for a future shortcode or whatever that does auto-numbered headings.)

Author: Oh yeah, good call.
